## SECTION A
1.Why are NumPy arrays preferred over Python lists?
- NumPy arrays are faster than normal Python lists when working with numbers. They use memory efficiently because they store elements in a fixed data type. NumPy also provides many built-in mathematical operations, which makes calculations easier in AI/ML. Therefore, NumPy arrays are commonly used for numerical calculations.

2.Explain vectorization. Why does replacing an explicit Python loop with a vectorized NumPy operation matter with large datasets? 
- Vectorization means performing an operation on the whole NumPy array without using a for loop. It makes the program shorter and usually faster. This is because NumPy performs the calculations using optimized code internally. Vectorization is very useful when working with large datasets

3.For compatible 2D arrays A and B, explain A * B vs A @ B. Describe one neural-network situation for each
- A * B means element-by-element multiplication, where corresponding elements are multiplied.
- A @ B means matrix multiplication, where rows and columns are multiplied and added.
- In neural networks, A * B can be used for element-wise operations, while A @ B is commonly used to multiply input data with weights.

4.What is broadcasting in NumPy? State the general rule used to decide whether arrays of different shapes can be combined element-wise.
- Broadcasting is a NumPy feature that allows arrays of different shapes to be used in an element-wise operation. NumPy automatically adjusts the smaller array to work with the larger array when their shapes are compatible. Two dimensions are compatible if they are equal or if one of them is 1. For example, an array of shape (3, 3) can be added to an array of shape (3, 1).


5.Explain flatten() vs ravel(). In a memory-constrained application processing large images, which would you prefer and why? 
- Both flatten() and ravel() are used to convert a multi-dimensional array into a one-dimensional array. flatten() creates a new copy of the array, while ravel() usually uses the original memory when possible. Therefore, in a memory-constrained application, I would prefer ravel() because it generally uses less extra memory.

6.For a 2D array, explain axis=0 and axis=1 in an aggregation function such as np.sum() or np.mean(). Illustrate with a small example. 
- For a 2D array, axis=0 means the operation is performed column-wise, while axis=1 means it is performed row-wise.
For example, for [[1, 2], [3, 4]], np.sum(array, axis=0) gives [4, 6] because 1+3=4 and 2+4=6.
np.sum(array, axis=1) gives [3, 7] because 1+2=3 and 3+4=7.

7.Why is it important to normalize or standardize numerical features before training most ML models? Explain normalization vs standardization. 
- Normalization or standardization puts numerical features into a similar range or scale. This is important because features with very large values can have more influence on some ML models. Normalization usually changes values to a fixed range such as 0 to 1, while standardization changes them so that the mean is 0 and standard deviation is 1. This can help the model train more effectively.

8.Why do functions such as np.nanmean() exist? What happens if np.mean() is used directly on data containing NaN?
- NaN means Not a Number and represents a missing or invalid numerical value. If we use np.mean() on data containing NaN, the result will usually also be NaN. np.nanmean() ignores the NaN values and calculates the mean using the available values. Therefore, it is useful when our dataset contains missing values.


9.In the ML convention X.shape = (samples, features), explain how this extends to a batch of RGB images. Write the exact shape for 64 RGB images of size 128 × 128. 
- In ML, X.shape = (samples, features) means the first value represents the number of data samples. For images, we can extend this idea by keeping the number of images first, followed by height, width, and color channels. For 64 RGB images of size 128 × 128, the exact shape is (64, 128, 128, 3). Here, 3 represents the RGB channels.


10.Why is reproducibility important during model development and experimentation? Refer to fixing a random seed with np.random.default_rng(seed). 
- Reproducibility means getting the same or very similar results when we run an experiment again. ML programs often use random values, so different runs can produce different results. We can fix the random values by using a seed, for example rng = np.random.default_rng(42). This makes experiments easier to test, compare, and debug.

# Section B

In [ ]:
# B1. Predict the exact output and/or resulting shape. Justify in one sentence. 
import numpy as np
a = np.array([1, 2, 3])
b = np.array([[1], [2], [3]])
print((a + b).shape)


(3, 3)


Explanation: a has shape (3,) and b has shape (3,1). NumPy broadcasts them together, producing a 3 × 3 array.

In [2]:
# B2. Predict the output and/or shape. [2 marks]
x = np.arange(1, 7).reshape(2, 3)
print(x[:, 1])


[2 5]


Explanation: : selects all rows and 1 selects the second column because indexing starts from 0. The resulting shape is (2,).

In [3]:
# B3. Predict the output and/or shape. [2 marks]
arr = np.array([5, 10, 15, 20, 25])
print(arr[arr % 10 == 0])


[10 20]


Explanation: arr % 10 == 0 checks which numbers are exactly divisible by 10. Therefore, only 10 and 20 are selected.

In [2]:
# B4. Predict the output and/or shape.
import numpy as np
m = np.array([[1, 2], [3, 4]])
print(m.T @ m)


[[10 14]
 [14 20]]


Explanation: m.T is the transpose of m, and @ performs matrix multiplication. The resulting shape is (2, 2).

In [3]:
# B5. Predict the output and/or shape. 
data = np.array([1, 2, np.nan, 4])

print(np.mean(data), np.nanmean(data))

nan 2.3333333333333335


Explanation: np.mean() returns nan because the array contains a NaN value. np.nanmean() ignores the NaN value and calculates the mean of 1, 2, 4, which is 7/3 = 2.3333.

In [4]:
# B6. Predict the output and/or shape. 
x = np.array([10, 20, 30])
y = np.where(x > 15, 1, 0)
print(y)

[0 1 1]


Explanation: np.where() returns 1 when the condition is true and 0 when it is false. Here, 20 and 30 are greater than 15, so their values become 1.


In [6]:
# B7. Predict the output and/or shape. 
a = np.zeros((3, 4))
print(a.ndim, a.size)

2 12


Explanation: The array has 2 dimensions because it is a 2D array. It contains 12 elements because 3 × 4 = 12.

In [7]:
# B8. Predict the output and/or shape. 
arr = np.arange(1, 13).reshape(3, 4)
print(arr[::-1, ::-1])


[[12 11 10  9]
 [ 8  7  6  5]
 [ 4  3  2  1]]


Explanation: [::-1] reverses the order of elements. The first [::-1] reverses the rows, and the second [::-1] reverses the columns. Therefore, the entire array is reversed.

## SECTION C

In [8]:
# C1. Given a 1D NumPy array of exam scores, use np.where() to produce labels 'P' (score >= 40) or 'F' (score < 40). 
scores = np.array([35, 45, 60, 28, 75, 39])
result = np.where(scores >= 40, 'P', 'F')
print(result)

['F' 'P' 'P' 'F' 'P' 'F']


In [9]:
# C2. Write a function standardize(x) that takes a 1D NumPy array and returns its z-score standardized version using only NumPy. 
def standardize(x):
    mean = np.mean(x)
    std = np.std(x)
    return (x - mean) / std
x = np.array([10, 20, 30, 40, 50])
result = standardize(x)
print(result)

[-1.41421356 -0.70710678  0.          0.70710678  1.41421356]


In [ ]:
# C3. Given two 2D NumPy arrays representing separate batches with the same number of features, combine them using an appropriate NumPy stacking function and print the resulting shape. 
batch1 = np.array([[1, 2, 3],
                   [4, 5, 6]])
batch2 = np.array([[7, 8, 9],
                   [10, 11, 12]])
combined = np.vstack((batch1, batch2))
print(combined)
print(combined.shape)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]
(4, 3)


In [12]:
# C4. Given X with shape (100, 5), compute column-wise mean and standard deviation and standardize the matrix in one vectorized expression. 
X = np.random.rand(100, 5)
mean = np.mean(X, axis=0)
std = np.std(X, axis=0)
X_standardized = (X - mean) / std
print(X_standardized.shape)

(100, 5)


In [13]:
# C5. Implement y_pred = X @ w + b for a feature matrix X, weight vector w and bias b. Then compute MSE using y_true. 
X = np.array([[1, 2],
              [2, 3],
              [3, 4]])
w = np.array([2, 3])
b = 1
y_true = np.array([9, 14, 19])
y_pred = X @ w + b
mse = np.mean((y_true - y_pred) ** 2)
print("Predicted values:", y_pred)
print("MSE:", mse)

Predicted values: [ 9 14 19]
MSE: 0.0


In [3]:
# C6. Using np.random.default_rng() with a fixed seed, generate 10 samples and 3 features from a normal distribution. Print the shape and confirm identical results with the same seed. 
rng1 = np.random.default_rng(42)
data1 = rng1.normal(size=(10, 3))
rng2 = np.random.default_rng(42)
data2 = rng2.normal(size=(10, 3))
print("Shape:", data1.shape)
print("Same results:", np.array_equal(data1, data2))

Shape: (10, 3)
Same results: True


In [4]:
# C7. Given a grayscale image as a 2D NumPy array with pixel values in [0, 255], normalize the values to [0, 1] using min-max normalization
image = np.array([[0, 100, 200],
                  [50, 150, 255]])
normalized = (image - image.min()) / (image.max() - image.min())
print(normalized)

[[0.         0.39215686 0.78431373]
 [0.19607843 0.58823529 1.        ]]


In [5]:
# C8. Given 4 RGB images of shape (28, 28, 3), stored as shape (4, 28, 28, 3), compute the average pixel value of each image across height, width and channels. State the axis argument and why. 
images = np.random.randint(0, 256, size=(4, 28, 28, 3))
average = np.mean(images, axis=(1, 2, 3))
print(average)
print(average.shape)

[125.87967687 127.86394558 128.61054422 129.0127551 ]
(4,)


## SECTION D


In [6]:
# D1. Data Preprocessing Pipeline
import numpy as np
data = np.array([10, 20, np.nan, 40, 50], dtype=float) # Sample 1D array containing NaN
print("Original data:")
print(data)

missing_values = np.isnan(data) # Step 1: Detect missing values
print("\nMissing value positions:")
print(missing_values)
print("Number of missing values:", np.isnan(data).sum())

mean_value = np.nanmean(data) # Step 2: Handle missing values using mean
print("\nMean value:", mean_value)
data[np.isnan(data)] = mean_value
print("Data after handling NaN:")
print(data)

mean = np.mean(data)  # Step 3: Standardize the clean data
std = np.std(data)
standardized_data = (data - mean) / std
print("\nMean after cleaning:", mean)
print("Standard deviation:", std)
print("\nStandardized data:")
print(standardized_data)

Original data:
[10. 20. nan 40. 50.]

Missing value positions:
[False False  True False False]
Number of missing values: 1

Mean value: 30.0
Data after handling NaN:
[10. 20. 30. 40. 50.]

Mean after cleaning: 30.0
Standard deviation: 14.142135623730951

Standardized data:
[-1.41421356 -0.70710678  0.          0.70710678  1.41421356]


Explanation

Data preprocessing means cleaning and preparing data before using it in a machine learning model.
First, detect NaN values using np.isnan(), replace them with the mean using np.nanmean(), and then standardize the cleaned data using (data - mean) / std.

D2. Image Batch Understanding 

1.Explain, with correct NumPy shapes, how a single RGB image differs from a batch of RGB images.
* A single RGB image of size 64 × 64 has the NumPy shape (64, 64, 3), where 3 represents Red, Green, and Blue channels.
* A batch of 16 such images has the shape (16, 64, 64, 3), where 16 represents the number of images.

In [7]:
# 2. Write code to create a dummy batch of 16 images of size 64 × 64 using random pixel values.
import numpy as np
images = np.random.randint(0, 256, size=(16, 64, 64, 3))
print("Batch shape:", images.shape)

Batch shape: (16, 64, 64, 3)


In [ ]:
# 3. Compute the average image of the batch, meaning the element-wise mean across all 16 images.
average_image = np.mean(images, axis=0)
print("Average image shape:", average_image.shape)

Average image shape: (64, 64, 3)


4.State the shape of your result and explain why.
- the result:
 Batch shape: (16, 64, 64, 3)
 Average image shape: (64, 64, 3)
- explaination:
 The result is (64, 64, 3) because axis=0 averages the corresponding pixels across all 16 images. The image height, width, and RGB channels remain unchanged.

D3. Debug the Model Code 

There are two main bugs in the given code.
- Bug 1: Incorrect shape of w

X has shape (3, 2), meaning it has 3 rows and 2 features.
Therefore, the weight vector must contain 2 weights, one for each feature.Its shape is (3,), which does not match the 2 columns of X.
- Bug 2: Incorrect operator precedence in MSE

The given expression is:mse = np.mean(y_true - y_pred ** 2) Python calculates the power ** before subtraction. Therefore, it actually calculates:
y_true - (y_pred ** 2)
But MSE should calculate the difference first, then square it:
mse = np.mean((y_true - y_pred) ** 2)

In [9]:
# correct code
X = np.array([[1, 2],
              [2, 3],
              [3, 4]])
w = np.array([0.5, 1.0])
b = 2.0
y_pred = X @ w + b
y_true = np.array([10, 20, 30])
mse = np.mean((y_true - y_pred) ** 2)
print("Predicted values:", y_pred)
print("MSE:", mse)

Predicted values: [4.5 6.  7.5]
MSE: 244.16666666666666


## SECTION E

In [10]:
# E1. Per-Column Min–Max Normalization 
X = np.array([
    [10, 100],
    [20, 200],
    [30, 300],
    [40, 400]
])
column_min = np.min(X, axis=0)  # minimum and maximum for each column
column_max = np.max(X, axis=0)

X_normalized = (X - column_min) / (column_max - column_min) # Per-column Min-Max normalization
print("Original matrix:")
print(X)

print("\nNormalized matrix:")
print(X_normalized)

Original matrix:
[[ 10 100]
 [ 20 200]
 [ 30 300]
 [ 40 400]]

Normalized matrix:
[[0.         0.        ]
 [0.33333333 0.33333333]
 [0.66666667 0.66666667]
 [1.         1.        ]]


In [11]:
# E2. Three-Level Score Classification 
scores = np.array([25, 40, 55, 70, 75, 90, 35])
result = np.where(
    scores < 40,
    "Fail",
    np.where(scores <= 70, "Average", "Excellent")
)
print("Scores:", scores)
print("Classification:", result)

Scores: [25 40 55 70 75 90 35]
Classification: ['Fail' 'Average' 'Average' 'Average' 'Excellent' 'Excellent' 'Fail']
